VirtoCommerce.Storefront.Model/Cart/LineItem.cs

```diff
@@ -312,12 +312,29 @@ public void ApplyRewards(IEnumerable<PromotionReward> rewards)
 
             DiscountAmount = new Money(Math.Max(0, (ListPrice - SalePrice).Amount), Currency);
 
+            if (Quantity == 0)
+            {
+                return;
+            }
+
             foreach (var reward in lineItemRewards)
             {
                 var discount = reward.ToDiscountModel(ListPrice - DiscountAmount);
                 if (reward.Quantity > 0)
                 {
-                    discount.Amount = discount.Amount * Math.Min(reward.Quantity, Quantity) / Quantity;
+                    //If we have condition like For "N in every Y …"
```

Comment:
You must encapsulate the implementation of  this new logic in the PromotionReward.ToDiscountModel() method. 
And please take a already exists implementation from here https://github.com/VirtoCommerce/vc-module-core/blob/dev/VirtoCommerce.Domain/Marketing/Model/Promotions/Rewards/AmountBasedReward.cs#L48

Reply:
Done. Please note that logic from vc-core-module AmountBasedReward does not contain storefront feature that MaxLimit works only on AmountType.Relative rewards (that's why i did not touch this method before - to not break existing storefront rewards logic for all items). Seems now it is not actual, and storefront calculation logic is the same as in vc-core-module AmountBasedReward. Also removed MarkertingTests that checked this AmountType.Relative feature.